workspaces/adventure-pack/goodies/typescript/Iterator.prototype.reduce/index.ts

```diff
@@ -0,0 +1,38 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  interface Iterator<T> {
+    reduce<U>(
+      this: Iterator<T>,
+      callbackFn: (accumulator: U, value: T, index: number) => U,
+      initialValue?: U,
+    ): U;
+  }
+}
+
+iteratorPrototype.reduce ??= function <T extends U, U>(
+  this: Iterator<T>,
+  callbackFn: (accumulator: U, value: T, index: number) => T,
+  initialValue?: U,
+): U {
+  let index = 0;
+  let accumulator: U = initialValue!;
+  let accumulatorInitialized = arguments.length > 1;
+
+  for (const element of this.toIterable()) {
```

Comment:
Yeah it's looking really good! 

For the detecting of T and U being the same type do you recommend making another method like [here](https://github.com/microsoft/TypeScript/blob/34048178a2b340518aa6baa888635788c9283055/src/lib/es5.d.ts#L1272)

Reply:
Nice find! It's not exactly another method, I believe it's more like multiple signatures for the same method, but yes, I think that's the way to do it.

Feel free to try it if you'd like, it can also be a separate PR.